aptos-move/aptos-vm/src/transaction_arg_validation.rs

```diff
@@ -83,22 +101,133 @@ pub(crate) fn validate_combine_signer_and_txn_args<S: MoveResolverExt>(
             .chain(args)
             .collect()
     };
+    if !needs_validation.is_empty() {
+        validate_args(session, &needs_validation, &combined_args, func)?;
+    }
     Ok(combined_args)
 }
 
-fn is_valid_txn_arg<S: MoveResolverExt>(session: &SessionExt<S>, typ: &Type) -> bool {
+// Return whether the argument is valid/allowed and whether it needs validation.
+// Validation is only needed for String arguments at the moment and vectors of them.
+fn is_valid_txn_arg<S: MoveResolverExt>(session: &SessionExt<S>, typ: &Type) -> (bool, bool) {
```

Comment:
this and all recursive related functions here are assumed to be "bounded" by the nested types in the signature allowed. 
I am not sure if that is a safe assumption or if we should put a limit to the recursion here

Reply:
Lets remember this for another PR. In generally we wanted to take a close look at recursion and bounds, but not know.